File: src/tools/RectangleTool.js
/**
 * MinecraftCAD - Rectangle Tool
 * Draws rectangular outlines or filled rectangles
 */

import { BaseTool } from './BaseTool.js';

export class RectangleTool extends BaseTool {
  constructor() {
    super('rectangle', 'Rectangle Tool');
    
    // Rectangle drawing state
    this.startPoint = null;
    this.endPoint = null;
    this.isDrawing = false;
    this.previewBlocks = [];
    this.fillMode = false; // Toggle between outline and filled
  }

  /**
   * Handle mouse down - start rectangle drawing
   */
  onMouseDown(event, worldPos) {
    if (!worldPos) return;

    if (!this.isDrawing) {
      // Start new rectangle
      this.startPoint = { ...worldPos };
      this.isDrawing = true;
      this.setCursor('crosshair');
    } else {
      // Complete rectangle
      this.endPoint = { ...worldPos };
      this.drawRectangle();
      this.finishDrawing();
    }
  }

  /**
   * Handle mouse move - update rectangle preview
   */
  onMouseMove(event, worldPos) {
    if (!this.isDrawing || !worldPos || !this.startPoint) return;

    this.endPoint = { ...worldPos };
    this.updatePreview();
  }

  /**
   * Handle right click - cancel current rectangle
   */
  onRightClick(event, worldPos) {
    if (this.isDrawing) {
      this.cancelDrawing();
    }
  }

  /**
   * Handle key press - F to toggle fill mode, ESC to cancel
   */
  onKeyDown(event) {
    if (event.key === 'Escape' && this.isDrawing) {
      this.cancelDrawing();
    } else if (event.key === 'f' || event.key === 'F') {
      this.fillMode = !this.fillMode;
      this.updatePreview(); // Refresh preview with new mode
      this.notifyStatusChange();
    }
  }

  /**
   * Draw the actual rectangle
   */
  drawRectangle() {
    if (!this.startPoint || !this.endPoint) return;

    const blocks = this.fillMode ? 
      this.calculateFilledRectangleBlocks(this.startPoint, this.endPoint) :
      this.calculateRectangleOutlineBlocks(this.startPoint, this.endPoint);
    
    // Block placement removed - tool disabled until new system is implemented
    console.log('Rectangle tool disabled - block placement system removed');

    // Clear preview
    this.clearPreview();
    
    // Trigger re-render
    this.notifyChange();
  }

  /**
   * Calculate blocks for rectangle outline
   */
  calculateRectangleOutlineBlocks(start, end) {
    const blocks = [];
    
    const minX = Math.min(start.x, end.x);
    const maxX = Math.max(start.x, end.x);
    const minY = Math.min(start.y, end.y);
    const maxY = Math.max(start.y, end.y);
    const minZ = Math.min(start.z, end.z);
    const maxZ = Math.max(start.z, end.z);

    // For 3D rectangles, we need to handle different planes
    if (minZ === maxZ) {
      // 2D rectangle on a single Z level
      // Top and bottom edges
      for (let x = minX; x <= maxX; x++) {
        blocks.push({ x, y: minY, z: minZ });
        if (minY !== maxY) {
          blocks.push({ x, y: maxY, z: minZ });
        }
      }
      
      // Left and right edges (excluding corners already added)
      for (let y = minY + 1; y < maxY; y++) {
        blocks.push({ x: minX, y, z: minZ });
        if (minX !== maxX) {
          blocks.push({ x: maxX, y, z: minZ });
        }
      }
    } else {
      // 3D rectangular frame - outline all faces
      this.addRectangleFrame(blocks, minX, maxX, minY, maxY, minZ, maxZ);
    }

    return blocks;
  }

  /**
   * Calculate blocks for filled rectangle
   */
  calculateFilledRectangleBlocks(start, end) {
    const blocks = [];
    
    const minX = Math.min(start.x, end.x);
    const maxX = Math.max(start.x, end.x);
    const minY = Math.min(start.y, end.y);
    const maxY = Math.max(start.y, end.y);
    const minZ = Math.min(start.z, end.z);
    const maxZ = Math.max(start.z, end.z);

    // Fill the entire rectangular volume
    for (let x = minX; x <= maxX; x++) {
      for (let y = minY; y <= maxY; y++) {
        for (let z = minZ; z <= maxZ; z++) {
          blocks.push({ x, y, z });
        }
      }
    }

    return blocks;
  }

  /**
   * Add rectangular frame blocks for 3D rectangles
   */
  addRectangleFrame(blocks, minX, maxX, minY, maxY, minZ, maxZ) {
    // Bottom face edges
    this.addRectangleEdges(blocks, minX, maxX, minY, maxY, minZ);
    
    // Top face edges (if different from bottom)
    if (minZ !== maxZ) {
      this.addRectangleEdges(blocks, minX, maxX, minY, maxY, maxZ);
    }
    
    // Vertical edges
    for (let z = minZ + 1; z < maxZ; z++) {
      blocks.push({ x: minX, y: minY, z });
      blocks.push({ x: maxX, y: minY, z });
      blocks.push({ x: minX, y: maxY, z });
      blocks.push({ x: maxX, y: maxY, z });
    }
  }

  /**
   * Add rectangle edges for a single Z level
   */
  addRectangleEdges(blocks, minX, maxX, minY, maxY, z) {
    // Top and bottom edges
    for (let x = minX; x <= maxX; x++) {
      blocks.push({ x, y: minY, z });
      if (minY !== maxY) {
        blocks.push({ x, y: maxY, z });
      }
    }
    
    // Left and right edges (excluding corners)
    for (let y = minY + 1; y < maxY; y++) {
      blocks.push({ x: minX, y, z });
      if (minX !== maxX) {
        blocks.push({ x: maxX, y, z });
      }
    }
  }

  /**
   * Update the rectangle preview
   */
  updatePreview() {
    this.clearPreview();
    
    if (!this.startPoint || !this.endPoint) return;

    this.previewBlocks = this.fillMode ? 
      this.calculateFilledRectangleBlocks(this.startPoint, this.endPoint) :
      this.calculateRectangleOutlineBlocks(this.startPoint, this.endPoint);
    
    // Add preview blocks to renderer (if supported)
    if (this.renderer && this.renderer.addPreviewBlocks) {
      this.renderer.addPreviewBlocks(this.previewBlocks, this.currentBlockType);
    }
  }

  /**
   * Clear preview blocks
   */
  clearPreview() {
    if (this.renderer && this.renderer.clearPreviewBlocks) {
      this.renderer.clearPreviewBlocks();
    }
    this.previewBlocks = [];
  }

  /**
   * Finish drawing and reset state
   */
  finishDrawing() {
    this.startPoint = null;
    this.endPoint = null;
    this.isDrawing = false;
    this.clearPreview();
    this.setCursor('default');
  }

  /**
   * Cancel current drawing operation
   */
  cancelDrawing() {
    this.finishDrawing();
  }

  /**
   * Activate the tool
   */
  activate() {
    super.activate();
    this.setCursor('crosshair');
  }

  /**
   * Deactivate the tool
   */
  deactivate() {
    this.cancelDrawing();
    super.deactivate();
  }

  /**
   * Get tool status text
   */
  getStatusText() {
    const modeText = this.fillMode ? 'filled' : 'outline';
    
    if (this.isDrawing) {
      return `Drawing ${modeText} rectangle - click to set corner, press F to toggle fill, right-click to cancel`;
    }
    return `Rectangle Tool (${modeText}) - click to start, press F to toggle fill mode`;
  }
}

File: src/tools/RectangleTool.lua
local BaseTool = require('src.tools.BaseTool')

local RectangleTool = setmetatable({}, BaseTool)
RectangleTool.__index = RectangleTool

function RectangleTool:new(blockData, appState)
    local self = setmetatable(BaseTool:new('rectangle', appState, blockData), RectangleTool)
    
    -- Ensure we have access to appState and blockData
    self.appState = appState
    self.blockData = blockData
    
    -- Rectangle drawing state
    self.isDrawing = false
    self.startX = 0
    self.startY = 0
    self.endX = 0
    self.endY = 0
    
    return self
end

function RectangleTool:onMousePressed(worldX, worldY, button)
    if button == 1 then -- Left click only
        -- Start drawing rectangle
        self.isDrawing = true
        self.startX, self.startY = self:snapToGrid(worldX, worldY)
        self.endX, self.endY = self.startX, self.startY
        
        return true
    end
    return false
end

function RectangleTool:onMouseReleased(worldX, worldY, button)
    if button == 1 and self.isDrawing then
        -- Finish drawing rectangle
        self.endX, self.endY = self:snapToGrid(worldX, worldY)
        
        -- Draw the rectangle
        local success = self:drawRectangle()
        
        -- Reset drawing state
        self.isDrawing = false
        
        return true
    end
    
    return false
end

function RectangleTool:onMouseMoved(worldX, worldY)
    if self.isDrawing then
        -- Update end position while dragging
        self.endX, self.endY = self:snapToGrid(worldX, worldY)
        return true
    end
end

function RectangleTool:drawRectangle()
    -- Get current level and block type
    local level = self.appState.currentLevel
    local blockType = self.appState.selectedBlockType
    
    if not blockType then
        return
    end
    
    if not level then
        return
    end
    
    -- Calculate rectangle bounds
    local minX = math.min(self.startX, self.endX)
    local maxX = math.max(self.startX, self.endX)
    local minY = math.min(self.startY, self.endY)
    local maxY = math.max(self.startY, self.endY)
    
    -- Place blocks to form rectangle outline
    local success = true
    local blocksPlaced = 0
    
    -- Top edge
    for x = minX, maxX do
        if self.blockData:setBlock(x, minY, level, blockType) then
            blocksPlaced = blocksPlaced + 1
        else
            success = false
        end
    end
    
    -- Bottom edge
    for x = minX, maxX do
        if self.blockData:setBlock(x, maxY, level, blockType) then
            blocksPlaced = blocksPlaced + 1
        else
            success = false
        end
    end
    
    -- Left edge
    for y = minY, maxY do
        if self.blockData:setBlock(minX, y, level, blockType) then
            blocksPlaced = blocksPlaced + 1
        else
            success = false
        end
    end
    
    -- Right edge
    for y = minY, maxY do
        if self.blockData:setBlock(maxX, y, level, blockType) then
            blocksPlaced = blocksPlaced + 1
        else
            success = false
        end
    end
    
    return success
end

function RectangleTool:snapToGrid(worldX, worldY)
    -- Snap to the center of the cell containing the mouse cursor
    local gridX = math.floor(worldX)
    local gridY = math.floor(worldY)
    
    return gridX, gridY
end

function RectangleTool:render()
    -- Draw preview rectangle while dragging
    if self.isDrawing then
        love.graphics.push()
        
        -- Set preview color (semi-transparent blue)
        love.graphics.setColor(0.392, 0.71, 0.965, 0.5)
        love.graphics.setLineWidth(2)
        
        -- Calculate rectangle bounds
        local minX = math.min(self.startX, self.endX)
        local maxX = math.max(self.startX, self.endX)
        local minY = math.min(self.startY, self.endY)
        local maxY = math.max(self.startY, self.endY)
        
        -- Draw rectangle outline
        local rectWidth = maxX - minX + 1
        local rectHeight = maxY - minY + 1
        
        -- Convert to screen coordinates (this is a simplified version)
        -- In a real implementation, you'd use the current renderer's worldToScreen method
        local screenX = minX * 32 -- Assuming 32px grid size
        local screenY = minY * 32
        
        love.graphics.rectangle('line', screenX, screenY, rectWidth * 32, rectHeight * 32)
        
        love.graphics.pop()
    end
end

return RectangleTool
